File: src/org/onion_lang/onion/compiler/util/MethodInvoker.java
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2005, Kota Mizushima, All rights reserved.       *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package org.onion_lang.onion.compiler.util;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

/**
 * @author Kota Mizushima
 * Date: 2005/09/15
 */
public class MethodInvoker {
  private MethodInvoker() {}
  
  public static Object call(Object target, String name, Object[] args)
    throws InvocationException {    
    try {
      return getMethod(target.getClass(), name, args).invoke(target, args);
    } catch (NoSuchMethodException e) {
      throw new InvocationException(e);
    } catch (IllegalArgumentException e) {
      throw new InvocationException(e);
    } catch (IllegalAccessException e) {
      throw new InvocationException(e);
    } catch (InvocationTargetException e) {
      throw new InvocationException(e);
    }
  }
  
  public static Object callStatic(Class target, String name, Object[] args) 
  	throws InvocationException{
    try {
      return getMethod(target, name, args).invoke(null, args);
    } catch (NoSuchMethodException e) {
      throw new InvocationException(e);
    } catch (IllegalArgumentException e) {
      throw new InvocationException(e);
    } catch (IllegalAccessException e) {
      throw new InvocationException(e);
    } catch (InvocationTargetException e) {
      throw new InvocationException(e);
    }

  }
  
  private static Method getMethod(Class target, String name, Object[] args)
  	throws NoSuchMethodException {
    Class[] argsClasses = new Class[args.length];
    for(int i = 0; i < args.length; i++) {
      argsClasses[i] = args[i].getClass();
    }
    return target.getMethod(name, argsClasses);    
  }
}

File: src/main/scala/onion/compiler/toolbox/MethodInvoker.scala
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2016-, Kota Mizushima, All rights reserved.  *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package onion.compiler.toolbox

import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method

/**
 * @author Kota Mizushima
 *
 */
object MethodInvoker {
  def call(target: AnyRef, name: String, args: Array[AnyRef]): AnyRef = {
    try {
      getMethod(target.getClass, name, args).invoke(target, args)
    } catch {
      case e: NoSuchMethodException =>
        throw new InvocationException(e)
      case e: IllegalArgumentException =>
        throw new InvocationException(e)
      case e: IllegalAccessException =>
        throw new InvocationException(e)
      case e: InvocationTargetException =>
        throw new InvocationException(e)
    }
  }

  def callStatic(target: Class[_], name: String, args: Array[AnyRef]): AnyRef = {
    try {
      getMethod(target, name, args).invoke(null, args)
    }
    catch {
      case e: NoSuchMethodException =>
        throw new InvocationException(e)
      case e: IllegalArgumentException =>
        throw new InvocationException(e)
      case e: IllegalAccessException =>
        throw new InvocationException(e)
      case e: InvocationTargetException =>
        throw new InvocationException(e)
    }
  }

  private def getMethod(target: Class[_], name: String, args: Array[AnyRef]): Method = {
    val argsClasses = args.map {_.getClass()}
    target.getMethod(name, argsClasses:_*)
  }
}
